Add proposal for Function and Values API

Awjin opened the pull request:
This proposal specifies the behavior of custom functions and Sass values in the new JS API.

Fixes https://github.com/sass/sass/issues/2510 and partially addresses #3056.

Awjin committed:
Draft string spec.

Awjin committed:
String API code review.

Awjin committed:
Add SassColor spec.

Awjin committed:
Code review.

- Be precise about JS / SassScript boundary.

Awjin committed:
Code review: colors and strings.

Awjin committed:
Add number spec.

Awjin committed:
--wip-- [skip ci]

Awjin committed:
Code review.

Awjin committed:
Add collections spec.

Awjin committed:
Code review.

Awjin committed:
Fine-tune wording.

Awjin committed:
Spec null and booleans.

Awjin committed:
Code review.

Awjin committed:
Add final bits

- Imports and modifies the new JS API.
- Background/summary.
- Fixes from last code review.

nex3 commented on proposal/new-function-and-values-api.d.ts:
```diff
@@ -0,0 +1,868 @@
+declare module './new-js-api' {
+  interface Options<sync extends 'sync' | 'async'> {
+    /**
+     * When the compiler encounters a function with a signature that does not
```
```suggestion
     * When the compiler encounters a global function call with a signature that does not
```

nex3 commented on proposal/new-function-and-values-api.d.ts:
```diff
@@ -0,0 +1,868 @@
+    /**
+     * When the compiler encounters a function with a signature that does not
+     * match that of a built-in function, but matches a key in this map, it must
+     * call the associated `CustomFunctionCallback`.
```
```suggestion
     * call the associated `CustomFunctionCallback` and return its result.
```

nex3 replied:
I think it's also worth explicitly saying "throw an error if this returns anything other than a `Value`". Normally we just say "failing to match type declarations produces undefined behavior", but this specific call feels likely enough to do something wrong (such as use an old-API function) that I think it's worth being explicit.

Awjin committed:
Add summary section.

Awjin committed:
Code review.

Awjin committed:
Make API use native Map.

The use of OrderedMap to enforce immutability should be an
implementation detail.

Awjin committed:
Revert "Make API use native Map."

This reverts commit 2244557ea6395353ddc1d9589e05ba37370f444d.

Awjin committed:
Make immutability airtight.

nex3 reviewed (approved):
Great work!
